File: scripts/CompareMvsPy/runEMforGMM.py
'''
Run EM algorithm for fitting Gaussian mixture model with K components

Author: Mike Hughes (mike@michaelchughes.com)

Usage
--------
Run as a script at the command line
>> python runEMforGMM.py datafilename K Niter savefilename seed

Arguments
--------
   datafilename  :  string path to .MAT file containing
                      observed data matrix X (each row is iid observation)

   K             :  integer # of components in mix model
   Niter         :  max # of iterations to run
   savefilename  :  string path to .MAT file to save final mixture model
                       with fields .w, .mu, and .Sigma
   seed          :  integer seed for random number generation (used for init only)
'''

import argparse
import time
import random
import numpy as np
import scipy.linalg
import scipy.io

CONVERGE_THR = 1e-6
MIN_COVAR    = 1e-8
EPS          = 1e-15

def parse_args():
  parser = argparse.ArgumentParser()
  parser.add_argument( 'datafilename' )
  parser.add_argument( 'K', type=int )
  parser.add_argument( 'Niter', type=int )
  parser.add_argument( 'savefilename', type=str)
  parser.add_argument( 'seed', type=int )
  parser.add_argument( '--MIN_COVAR', type=float, default=MIN_COVAR )
  return parser.parse_args()

def main( datafilename, K, Niter, savefilename, seed ):
  X = scipy.io.loadmat( datafilename )['X']
  loglik = -np.inf*np.ones( Niter )

  Resp = init_responsibilities( X, K, seed)
  tstart = time.time()
  for t in xrange( Niter ):
    model = Mstep( X, Resp )
    Resp, loglik[t] = Estep( X, model )
    
    print '%5d/%d after %.0f sec | %.8e' % (t+1, Niter, time.time()-tstart, loglik[t])    
    deltaLogLik = loglik[t] - loglik[t-1]
    if deltaLogLik < CONVERGE_THR:
      break
    if deltaLogLik < 0:
      print 'WARNING: loglik decreased!'
  scipy.io.savemat( savefilename, model, oned_as='row' ) # oned_as kwarg avoid stupid warning
  return model, loglik

def init_responsibilities( X, K, seed):
  N,D = X.shape
  random.seed( seed )
  rowIDs = random.sample( xrange(N), K ) #without replacement
  mu = X[rowIDs, : ]

  logResp = np.zeros( (N, K) )
  for k in xrange( K ):
    logResp[:,k] = loggausspdf( X, mu[k,:], np.eye(D) )
  logPrPerRow = logsumexp( logResp, axis=1 )
  Resp = np.exp( logResp - logPrPerRow[:,np.newaxis] )
  return Resp

###################################################
def Estep(X, model):
  w = model['w']
  mu = model['mu']
  Sigma = model['Sigma']

  N = X.shape[0]
  K = mu.shape[0]
  logResp = np.zeros( (N, K) )
  for k in xrange( K ):
    logResp[:,k] = loggausspdf( X, mu[k,:], Sigma[:,:,k] )
  logResp += np.log( w )

  logPrPerRow = logsumexp( logResp, axis=1 )
  Resp = np.exp( logResp - logPrPerRow[:,np.newaxis] )
  return Resp, np.sum(logPrPerRow)
  
def Mstep(X, Resp):
  N,D = X.shape
  K = Resp.shape[1]

  Nk = np.sum( Resp, axis=0) + EPS
  w  = Nk/N
  mu = np.dot( Resp.T, X ) / Nk[:,np.newaxis]
  Sigma = np.zeros( (D,D,K) )
  for k in xrange( K ):
    Xdiff = X - mu[k,:]
    Xdiff = Xdiff * np.sqrt( Resp[:,k] )[:,np.newaxis]
    Sigma[:,:,k] = np.dot( Xdiff.T, Xdiff) / Nk[k] + MIN_COVAR*np.eye(D)
  return dict( w=w, mu=mu, Sigma=Sigma )

###################################################
def loggausspdf( X, mu, Sigma):
  ''' Calc log p( x | mu, Sigma) for each row of matrix X
  '''
  N,D = X.shape
  dist, cholSigma = distMahal( X, mu, Sigma )
  logdetSigma = 2*np.sum( np.log( np.diag(cholSigma) ) )
  logNormConst = -0.5*D*np.log(2*np.pi) - 0.5*logdetSigma
  logpdfPerRow = logNormConst - 0.5*dist
  return logpdfPerRow
  
def distMahal( X, mu, Sigma ):
  ''' Calc mahalanobis distance: (x-mu)^T Sigma^{-1} (x-mu)
       for each row of matrix X
  '''
  N,D = X.shape
  Xdiff = X - mu
  cholSigma = scipy.linalg.cholesky( Sigma, lower=True)
  Q = scipy.linalg.solve_triangular( cholSigma, Xdiff.T, lower=True )
  distPerRow = np.sum( Q**2, axis=0 )
  return distPerRow, cholSigma


def logsumexp( logA, axis=None):
  ''' Compute log( sum(exp(logA))) in numerically stable way
  '''
  logA = np.asarray( logA )
  logAmax = logA.max( axis=axis )

  if axis is None:
    logA = logA - logAmax
  elif axis==1:
    logA = logA - logAmax[:,np.newaxis]
  elif axis==0:
    logA = logA - logAmax[np.newaxis,:]
  assert np.allclose( logA.max(), 0.0 )
  logA = np.log( np.sum( np.exp( logA ), axis=axis )  )
  return logA + logAmax


if __name__ == '__main__':
  args = parse_args()
  MIN_COVAR = args.MIN_COVAR
  main(  args.datafilename, args.K, args.Niter, args.savefilename, args.seed )

File: scripts/CompareMvsPy/runEMforGMM.m
function [model,loglik] = runEMforGMM( data, K, Niter, savefilename, seed)
% Run EM algorithm for fitting the Gaussian mixture model with K components
% AUTHOR: Mike Hughes (mike@michaelchughes.com)
% INPUT
%   data  : NxD data matrix (each row is one i.i.d. observation)
%           or string path to .mat file
%   K     : # of mixture components (integer)
%   Niter : # of iterations to run
%   savefilename : string path to .mat file to save model
%   seed  : integer seed for random number generation (used for init only)
% ACKNOWLEDGEMENTS
%  Based on Matlab Central implementation by Michael (Mo) Chen.

if ischar( data )
    X = load( data );
    X = X.X;
else
    X = data;
end

CONVERGE_THR = 1e-6;
loglik = -inf( 1, Niter );

fprintf( 'EM for Mixture of %d Gaussians | seed=%d\n', K, seed);
Resp = init_responsibilities( X, K, seed );
tic;
for t = 1:Niter
    model = Mstep( X, Resp );
    [Resp, loglik(t)] = Estep( X, model );
    
    fprintf( ' %5d/%d after %.0f sec | %.8e \n', t, Niter, toc, loglik(t)  );
    
    if t > 1
        deltaLogLik = loglik(t) - loglik(t-1);
        if deltaLogLik < 0
            fprintf('WARNING: log lik decreased!\n');
        end
        if deltaLogLik < CONVERGE_THR
            break;
        end
    end
end
loglik = loglik( ~isinf(loglik) );
save( savefilename, '-struct', 'model' );
end

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
function Resp = init_responsibilities(X, K, seed)
[N,D] = size(X);
if true    
    % Use python to generate random samples, so we can compare results
    PYCMD = sprintf( 'python -c "import random; random.seed(%d); print random.sample(xrange(%d),%d)"', seed, N, K);
    [status, resultStr] = system( PYCMD );
    rowIDs = str2num( resultStr );
    mu = X( rowIDs+1, :); % python output is 0-idx, so convert to matlab
else
    % Use matlab to generate random samples
    doReplace = false;
    rng(seed);
    rowIDs = randsample( N, K, doReplace );
    mu = X(rowIDs,:);
end
logResp = zeros(N,K);
for k = 1:K
    logResp(:,k) = loggausspdf( X, mu(k,:), eye(D) );
end
logProbPerRow = logsumexp(logResp,2);
Resp = exp( bsxfun(@minus,logResp, logProbPerRow) );
end

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
function [Resp, loglik] = Estep( X, model)
w = model.w;
mu = model.mu;
Sigma = model.Sigma;

N = size(X,1);
K = size(mu,1);

logResp = zeros(N,K);
for k = 1:K
    logResp(:,k) = loggausspdf(X, mu(k,:), Sigma(:,:,k) );
end
logResp = bsxfun(@plus, logResp, log(w));
logProbPerRow = logsumexp(logResp,2);

Resp = exp( bsxfun(@minus,logResp, logProbPerRow) );
loglik = sum( logProbPerRow );
end

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
function model = Mstep(X, Resp)
[N,D] = size(X);
K = size(Resp,2);

Nk = sum( Resp,1) + eps;
w = Nk/N;
mu = bsxfun(@times, Resp'*X, 1./Nk' );
Sigma = zeros(D,D,K);
for k = 1:K
    Xdiff = bsxfun( @minus, X, mu(k,:) );
    Xdiff = bsxfun( @times, Xdiff, sqrt(Resp(:,k)) );
    Sigma(:,:,k) = (Xdiff'*Xdiff)/Nk(k) + 1e-8*eye(D);
end
model.w = w;
model.mu = mu;
model.Sigma = Sigma;
end

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
function logpdf = loggausspdf(X, mu, Sigma)
% Compute log pdf for each row in matrix X
%   logpdf(n) = log Pr( X(n,:) | mu, Sigma )

[N,D] = size(X);
[distM,cholSigma] = distMahal( X, mu, Sigma );
logdetSigma = 2*sum( log( diag( cholSigma )));
logNormConst = -0.5*D*log(2*pi) - 0.5*logdetSigma;
logpdf = logNormConst - 0.5*distM;
end

function [distPerRow, cholSigma] = distMahal( X, mu, Sigma )
[N,D] = size(X);
Xdiff = bsxfun(@minus,X,mu);
[cholSigma, status]= chol(Sigma, 'lower');
if status ~= 0
    error('ERROR: Sigma is not PD.');
end
Q = cholSigma\Xdiff';
distPerRow = sum( Q.^2, 1);  % quadratic term (M distance)
end

function logS = logsumexp( logA, dim)
% Compute log(sum(exp(logA),dim)) while avoiding numerical underflow.
logAMax = max( logA,[],dim);
logA = bsxfun(@minus,logA,logAMax);
logS = logAMax + log(sum(exp(logA),dim));
end
